### ISSUE TITLE
nn.CrossEntropyLoss() yields wrong output for big logits

### ISSUE THREAD

**Issue by fab-jul** [pytorch/pytorch#11752]
## Issue description

`nn.CrossEntropyLoss()` yields incorrect results for big logits.

## Code example

``` python
import torch
from torch import nn

crit = nn.CrossEntropyLoss()

target = torch.ones(1, dtype=torch.long)

y_small = torch.ones(1, 2, dtype=torch.float32)
y_big = y_small * 1e8

loss_small = crit(y_small, target)
loss_big = crit(y_big, target)
loss_fixed = crit(y_big - y_big.max(), target)

print('loss_small = {}\n'
      'loss_big   = {}\n'
      'loss_fixed = {}'.format(
        loss_small, loss_big, loss_fixed))
```

`loss_small` should be equal to `loss_big`, as both `y_small` and `y_big` contain logits that correspond to the same distribution. However, the output of the above is:

```
loss_small = 0.6931471824645996
loss_big   = 0.0
loss_fixed = 0.6931471824645996
```

Note that the output is correct (`loss_fixed == loss_small`) if I apply the trick `softmax(x) = softmax(x - c)` for some `c`, in particular, for `c=max(x)`.

This happens on CPU and GPU.

## System Info

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Debian GNU/Linux 8 (jessie)
GCC version: (Debian 4.9.2-10+deb8u1) 4.9.2
CMake version: version 3.0.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: 
GPU 0: TITAN Xp

Nvidia driver version: 384.130
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy (1.14.3)
[pip] torch (0.4.1)
[pip] torchvision (0.2.1)
[conda] pytorch                   0.4.1           py36_cuda9.0.176_cudnn7.1.2_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

**Comment by fmassa** [comment]
I think this boils down to `log_softmax`.

```python
y_small = torch.ones(1, 2, dtype=torch.float32)
y_big = y_small * 1e8
print(torch.nn.functional.log_softmax(y_small, -1))
print(torch.nn.functional.log_softmax(y_big, -1))
```
yields
```python
tensor([[-0.6931, -0.6931]])
tensor([[0., 0.]])
```

I'm not yet sure where the problem is (I couldn't find anything suspicious from a quick look at [the CPU code](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L44-L60)), but for information it works as expected if we cast the tensors to double precision
```python
print(torch.nn.functional.log_softmax(y_small.double(), -1))
print(torch.nn.functional.log_softmax(y_big.double(), -1))
```
gives
```python
tensor([[-0.6931, -0.6931]], dtype=torch.float64)
tensor([[-0.6931, -0.6931]], dtype=torch.float64)
```

**Comment by vadimkantorov** [comment]
`max_input` [seems to be computed](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L41-L42) across dimensions for every spatial location. Can this cause problems?

**Comment by fab-jul** [comment]
Good insight, however, for double I get the same fail for `1e8 ** 2 == 1e16`:

```python
y_small = torch.ones(1, 2, dtype=torch.double)
y_big = y_small * 1e16
print(torch.nn.functional.log_softmax(y_small, -1))
print(torch.nn.functional.log_softmax(y_big, -1))

# output
tensor([[-0.6931, -0.6931]], dtype=torch.float64)
tensor([[0., 0.]], dtype=torch.float64)
```

**Comment by sytrus-in-github** [comment]
I think there is a numerical problem in [the CPU code](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L44-L60) especially at [line 54](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SoftMax.cpp#L54):

In this case `input_data[d * dim_stride] = max_input = 1e8` and `z = 1`. `tmpsum = 2` before line 54.

At line 54, we have `tmpsum = 1e8 + log(2) = 1e8`. Numerically with float precision the `log(2)` is ignored so at the end we basically have ` 1e8 - (1e8 + log(2)) = 0` instead of `-log(2)`.

Changing line 54 to
```python
tmpsum = std::log(tmpsum);
```
and line 60 to
```python
output_data[d * dim_stride] = input_data[d * dim_stride] - max_input - tmpsum;
```
should hopefully solve the problem. (and also the case with double precision)

I am quite busy right now. But if these are the only lines that need to change. I don't mind do a PR :)

**Comment by soumith** [comment]
fixed via https://github.com/pytorch/pytorch/pull/11866#pullrequestreview-156960313 thanks to @sytrus-in-github (landing in a couple of hours)
